packages/frontend/app/services/channels/channel-verifier.ts

```diff
@@ -0,0 +1,131 @@
+import Service, { inject as service } from '@ember/service';
+import ChannelContextChain from 'emberclear/models/channel-context-chain';
+import VoteChain, { VOTE_ACTION } from 'emberclear/models/vote-chain';
+import VoteVerifier from 'emberclear/services/channels/vote-verifier';
+import Identity from 'emberclear/models/identity';
+import { identityEquals, identityArrayContains } from 'emberclear/utils/identity-comparison';
+
+export default class ChannelVerifier extends Service {
+  @service('channels/vote-verifier') voteVerifier!: VoteVerifier;
+
+  async verify(channel: ChannelContextChain): Promise<boolean> {
+    if (channel.previousChain === undefined) {
+      return this.singleChainVerifier(channel);
+    }
+    if (
+      !(await this.verify(channel.previousChain)) ||
+      !(await this.voteVerifier.verify(channel.supportingVote)) ||
+      !this.checkVoteCompletedPositive(channel.supportingVote, channel.previousChain.admin)
+    ) {
+      return false;
+    }
+
+    switch (channel.supportingVote.action) {
+      case VOTE_ACTION.ADD:
+        return this.checkAdd(channel);
+      case VOTE_ACTION.PROMOTE:
+        return this.checkPromote(channel);
+      case VOTE_ACTION.REMOVE:
+        return this.checkRemove(channel);
+      default:
+        return false;
+    }
+  }
+
+  private singleChainVerifier(channel: ChannelContextChain): boolean {
+    if (
```

Comment:
this is somewhat styleistic, but also may read nicer -- if the contents of the if condition were extracted to a variable, instead of returning true / false, you could just return the variable:

example:

```ts
let isInvalid = ( big long conditional )

return !isInvalid;
```

Reply:
addressed in https://github.com/NullVoxPopuli/emberclear/pull/738/commits/d4cba3fd1da38e46cb82af774a584a0f014ed56d